cooked-validators/src/Cooked/MockChain/Monad.hs

```diff
@@ -105,6 +109,41 @@ spendableRef txORef = do
   Just txOut <- txOutByRef txORef
   return (txORef, fromJust (Pl.fromTxOut txOut))
 
+-- | Some values of type "SpendableOut" have no explicit datum in their
+-- "ChainIndexTxOut" but the datum hash instead. When used in cooked
+-- constraints, this results in a runtime error. This function modifies the
+-- value to replace the datum hash by the datum by looking it up in the state
+-- of the block chain.
+spOutResolveDatum ::
+  MonadBlockChain m =>
+  SpendableOut ->
+  m SpendableOut
+spOutResolveDatum (txOutRef, chainIndexTxOut@(Pl.ScriptChainIndexTxOut _ _ (Left _) _)) = do
+  mDatum <- datumFromTxOut chainIndexTxOut
+  case mDatum of
+    Nothing -> fail "datum hash not found in block chain state"
+    Just datum ->
+      return
+        (txOutRef, chainIndexTxOut {Pl._ciTxOutDatum = Right datum})
+spOutResolveDatum spOut = return spOut
+
+-- | Retrieve the ordered list of "SpendableOutput" corresponding to each
+-- output of the given "CardanoTx". These "SpendableOutput" are processed to
+-- remove unresolved datum hashes with "spOutResolveDatum" so that they can be
+-- used in spending constraints.
+--
+-- This is useful when writing endpoints and/or traces to fetch utxos of
+-- interest right from the start and avoid querying the chain for them
+-- afterwards using "utxosSuchThat" functions.
+spOutsFromCardanoTx :: MonadBlockChain m => Pl.CardanoTx -> m [SpendableOut]
+spOutsFromCardanoTx cardanoTx = forM (Pl.getCardanoTxOutRefs cardanoTx) go
+  where
+    go :: MonadBlockChain m => (Pl.TxOut, Pl.TxOutRef) -> m SpendableOut
+    go (txOut, txOutRef) =
+      case Pl.fromTxOut txOut of
+        Just chainIndexTxOut -> spOutResolveDatum (txOutRef, chainIndexTxOut)
+        Nothing -> fail "could not extract ChainIndexTxOut"
+
```

Comment:
Maybe
```haskell
spOutsFromCardanoTx cardanoTx =
    forM (Pl.getCardanoTxOutRefs cardanoTx) $ \(txOut, txOutRef) ->
      case Pl.fromTxOut txOut of
        Just chainIndexTxOut -> spOutResolveDatum (txOutRef, chainIndexTxOut)
        Nothing -> fail "could not extract ChainIndexTxOut"
```
The `go` function is more often used for loops than for the actual body of the iteration.

Another option could be renaming `go` with `toSpendableOut`.